iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp

```diff
@@ -38,19 +44,23 @@ class ServiceDiscovery
     /// @param[in] service service string to search for, a nullopt corresponds to a wildcard
     /// @param[in] instance instance string to search for, a nullopt corresponds to a wildcard
     /// @param[in] event event string to search for, a nullopt corresponds to a wildcard
-    /// @return cxx::expected<ServiceContainer, FindServiceError>
-    /// ServiceContainer: on success, container that is filled with all matching instances
-    /// FindServiceError: if any, encountered during the operation
-    cxx::expected<ServiceContainer, FindServiceError>
-    findService(const cxx::optional<capro::IdString_t>& service,
-                const cxx::optional<capro::IdString_t>& instance,
-                const cxx::optional<capro::IdString_t>& event) noexcept;
-
-    /// @brief requests the serviceRegistryChangeCounter from the shared memory
-    /// @return pointer to the serviceRegistryChangeCounter
-    /// @todo #415 If this method is still used after refactoring, consider to return a reference so that a nullptr
-    /// check is not necessary
-    virtual const std::atomic<uint64_t>* getServiceRegistryChangeCounter() noexcept;
+    /// @return ServiceContainer
+    /// ServiceContainer: container that is filled with all matching instances
+    ServiceContainer findService(const cxx::optional<capro::IdString_t>& service,
+                                 const cxx::optional<capro::IdString_t>& instance,
+                                 const cxx::optional<capro::IdString_t>& event) noexcept;
+
+    friend iox::popo::NotificationAttorney;
+
+  private:
+    void enableEvent(popo::TriggerHandle&& triggerHandle, const ServiceDiscoveryEvent event) noexcept;
+    void disableEvent(const ServiceDiscoveryEvent event) noexcept;
+    void invalidateTrigger(const uint64_t uniqueTriggerId);
+    iox::popo::WaitSetIsConditionSatisfiedCallback
+    getCallbackForIsStateConditionSatisfied(const popo::SubscriberState state);
+    roudi::ServiceRegistry m_serviceRegistry;
+    popo::Subscriber<roudi::ServiceRegistry> m_serviceRegistrySubscriber{
+        roudi::serviceRegistryService, {1U, 1U, iox::NodeName_t("Service Registry"), true}};
```

Comment:
Please create constexpr variables for all arguments. 
Why do we need a node name here? Aren't nodes supposed to be unique but with this approach we would have a `Service Registry` node in every process?!

Reply:
Refactored the strings.

The node name can be handy to see that the different internal publishers in RouDi belong to different modules or in case of an app visually show the users that this is an special subscriber. No, uniqueness comes in place here, it's just a grouping based on names.